File: civitas/src/civitas/Jugador.java
package civitas;

import static java.lang.Float.compare;
import java.util.ArrayList;
import GUI.Dado;

/**
 *
 * 
 */
public class Jugador implements Comparable<Jugador>{
   
    static protected int CasasMax = 4;
    static protected int CasasPorHotel = 4;
    protected boolean encarcelado;    
    static protected int HotelesMax = 4;   
    private String nombre;   
    private int numCasillaActual;   
    static protected float PasoPorSalida = 1000f;    
    static protected float PrecioLibertad = 200f;    
    private boolean puedeComprar;    
    protected float saldo;    
    static float SaldoInicial = 7500f;
    protected SorpresaSalirCarcel salvoconducto;    
    protected ArrayList<TituloPropiedad> propiedades;
    protected  boolean especulador;
    
    
    /**
     * Cancela la hipoteca de una propiedad.
     * @param ip
     * @return 
     */
    boolean cancelarHipoteca(int ip) {
        boolean result = false;
        if(encarcelado) {
            return result;
        }
        if(existeLaPropiedad(ip)) {
            TituloPropiedad propiedad = propiedades.get(ip);
            float cantidad = propiedad.getImporteCancelarHipoteca();
            boolean puedoGastar = puedoGastar(cantidad);
            if(puedoGastar) {
                result = propiedad.cancelarHipoteca(this);
                if(result) {
                    Diario.getInstance().ocurreEvento("El jugador " + nombre +
                            " cancela la hipoteca de la propiedad " + ip);
                }
            }
        }
        return result;
    }
    
    /**
     * Calcula el numero de casas mas hoteles.
     * @return cantidad.
     */
    int cantidadCasasHoteles() {
        int cantidad = 0;
        for(int i = 0; i < propiedades.size(); i++) {
            cantidad += propiedades.get(i).cantidadCasasHoteles();
        }
        return cantidad;
    }
    
    /**
     * Compara dos jugadores por su saldo.
     * @param otro
     * @return 
     */
    @Override
    public int compareTo(Jugador otro) {
        return compare(saldo, otro.getSaldo());
    }
    
    /**
     * Compra una propiedad.
     * @param titulo : de la propiedad.
     * @return TRUE si se ha comprado, FALSE si no.
     */
    boolean comprar(TituloPropiedad titulo) {
        boolean result = false;
        if(encarcelado) {
            return result; 
        }
        if(puedeComprar) {
            float precio = titulo.getPrecioCompra();
            if(puedoGastar(precio)) {
                result = titulo.comprar(this);
                if(result) {
                    propiedades.add(titulo);
                    Diario.getInstance().ocurreEvento("El jugador " + nombre + 
                            " ha comprado la propiedad " + titulo.toString());                    
                }
                puedeComprar = false;
            }
        }
        return result;
    }
    
    /**
     * Construye una casa en la propiedad actual.
     * @param ip : identificador de la propiedad.
     * @return TRUE si se construye la casa, FALSE si no.
     */
    boolean construirCasa(int ip) {                                
        boolean result = false;
        boolean puedoEdificarCasa;
        if(encarcelado){
            return result;
        }else{
            if(this.existeLaPropiedad(ip)) {
                TituloPropiedad propiedad = propiedades.get(ip);
                puedoEdificarCasa = this.puedoEdificarCasa(propiedad);
                float precio = propiedad.getPrecioEdificar();
                if(puedoGastar(precio) && propiedad.getNumCasas() < getCasasMax()) {
                    puedoEdificarCasa = true;
                }
                if(puedoEdificarCasa) {
                    result = propiedad.construirCasa(this);
                    if(result) {
                        Diario.getInstance().ocurreEvento("El jugador " + nombre
                                + " construye una casa en la propiedad " + 
                                propiedades.get(ip).getNombre());
                    }
                } 
            }
        }
        return result;
    }
    
    /**
     * Construye un hotel en la propiedad actual.
     * @param ip : identificador de la propiedad.
     * @return TRUE si se construye el hotel, FALSE si no.
     */
    boolean construirHotel(int ip) {
        boolean result = false;
        if(encarcelado) {
            return result;
        }
        if(this.existeLaPropiedad(ip)) {
            TituloPropiedad propiedad = propiedades.get(ip);
            boolean puedoEdificarHotel = this.puedoEdificarHotel(propiedad);
            float precio = propiedad.getPrecioEdificar();
            if(puedoGastar(precio) && (propiedad.getNumHoteles() < 
                    this.getHotelesMax()) && (propiedad.getNumCasas() >= 
                    this.getCasasPorHotel())) {
                puedoEdificarHotel = true;
            }
            if(puedoEdificarHotel) {
                result = propiedad.construirHotel(this);
                propiedad.derruirCasas(CasasPorHotel, this);
                Diario.getInstance().ocurreEvento("El jugador " + nombre + 
                        " construye un hotel en la propiedad " + 
                        propiedades.get(ip).getNombre());
            }
        }
        return result;
    }
    
    /**
     * Comprueba si el jugador debe ser encarcelado.
     * @return TRUE si el jugador debe ser encarcelado.
     */
    protected boolean debeSerEncarcelado() {
        if(isEncarcelado()) {
            return false;
        }else{
            if(!tieneSalvoconducto()) {
                return true;
            }else{
                perderSalvoconducto();
                Diario.getInstance().ocurreEvento("El jugador " + nombre + 
                        " se libra de la carcel");
                return false;
            }    
        }
    }
    
    /**
     * Comprueba si el jugador esta en bancarrota.
     * @return TRUE si el jugador esta en bancarrota.
     */
    boolean enBancarrota() {
        return saldo <= 0;
    }
    
    /**
     * Encarcela al jugador si debe ser encarcelado y se informa.
     * @param numCasillaCarcel casilla de la carcel.
     * @return TRUE si se encarcela, FALSE si no.
     */
    boolean encarcelar(int numCasillaCarcel) {
        if(debeSerEncarcelado()) {
            moverACasilla(numCasillaCarcel);
            encarcelado = true;
            Diario.getInstance().ocurreEvento("El jugador " + nombre 
                    + " ha sido encarcelado");
        }
        return encarcelado;
    }

    /**
     * Comprueba si existe la propiedad.
     * @param ip indice de la propiedad.
     * @return TRUE si existe, FALSE si no.
     */
    protected boolean existeLaPropiedad(int ip) {
        return ip <= propiedades.size();
    }
    
    /**
     * Consulta el maximo de casas.
     * @return CasasMax.
     */
    protected int getCasasMax() {
        return CasasMax;
    }
    
    /**
     * Consulta el numero de casas por hotel.
     * @return CasasPorHotel.
     */
    int getCasasPorHotel() {
        return CasasPorHotel;
    }
    
    public boolean getEspeculador() {
        return especulador;
    }
    
    /**
     * Consulta el numero maximo de hoteles.
     * @return HotelesMax.
     */
    protected int getHotelesMax() {
        return HotelesMax;
    }
    
    /**
     * Consulta el nombre de la propiedad.
     * @return nombre de la propiedad.
     */
    public String getNombre() {
        return nombre;
    }
    
    /**
     * Consulta el nombre de las propiedades
     * @return 
     */
    public ArrayList<String> getNombrePropiedades(){
       ArrayList<String> nombres = new ArrayList<>();
       for(int i = 0; i < propiedades.size(); i++) {
           nombres.add(propiedades.get(i).getNombre());
       }
       return nombres;
    }
    
    
    /**
     * Consulta la casilla actual.
     * @return casilla actual.
     */
    int getNumCasillaActual() {
        return numCasillaActual;
    }
    
    /**
     * Consulta el precio a pagar para salir libre.
     * @return PrecioLibertad.
     */
    private float getPrecioLibertad() {
        return PrecioLibertad;
    }
    
    /**
     * Consulta el precio del paso por salida.
     * @return PasoPorSalida.
     */
    private float getPremioPasoSalida() {
        return PasoPorSalida;
    }
    
    /**
     * Consulta si tiene salvoconducto.
     * @return salvoconducto.
     */
    Sorpresa getSalvoconducto(){
        return salvoconducto;
    }
    
    /**
     * Consulta las propiedades del jugador.
     * @return propiedades.
     */
    public ArrayList<TituloPropiedad> getPropiedades() {
        return propiedades;
    }
    
    /**
     * Consulta si se puede comprar la propiedad.
     * @return TRUE si puede, FALSE si no.
     */
    boolean getPuedeComprar() {
        return puedeComprar;
    }
    
    /**
     * Consulta el saldo del jugador.
     * @return saldo.
     */
    public float getSaldo() {
        return saldo;
    }
    
    /**
     * Hipoteca la propiedad del jugador.
     * @param ip : identificador de la propiedad.
     * @return TRUE si hipoteca, FALSE si no.
     */
    boolean hipotecar(int ip) {
        boolean resultado = false;
        if(encarcelado) {
            return resultado;
        }
        if(existeLaPropiedad(ip)) {
            TituloPropiedad propiedad = propiedades.get(ip);
            resultado = propiedad.hipotecar(this);
        }
        if(resultado) {
            Diario.getInstance().ocurreEvento("El jugador " + nombre +
                    " hipoteca la propiedad " + ip);
        }
        return resultado;
    }
    
    /**
     * Comprueba si el jugador esta encarcelado.
     * @return TRUE si esta encarcelado, FALSE si no.
     */
    public boolean isEncarcelado() {
        return encarcelado;
    }
    
    /**
     * Constructor que recibe un nombre.
     * @param nombre 
     */
    Jugador(String nombre) {
        this.nombre = nombre;
        numCasillaActual = 0;
        saldo = 7500;
        puedeComprar = true;
        propiedades = new ArrayList<>();
        encarcelado = false;
        salvoconducto = null;
        especulador = false;
    }
    
    /**
     * Constructor que copia un jugador en otro.
     * @param otro 
     */
    protected Jugador(Jugador otro) {
        this.nombre = otro.nombre;
        this.encarcelado = otro.encarcelado;
        this.numCasillaActual = otro.numCasillaActual;
        this.puedeComprar = otro.puedeComprar;
        this.saldo = otro.saldo;
        this.salvoconducto = otro.salvoconducto;
        this.propiedades = otro.propiedades;
        this.especulador = otro.especulador;
    }
    
    /**
     * Modifica el saldo del jugador sumandole una cantidad.
     * @param cantidad
     * @return TRUE si se ha modificado, FALSE si no.
     */
    boolean modificarSaldo(float cantidad) {
        this.saldo += cantidad;
        Diario.getInstance().ocurreEvento("Se ha modificado el saldo de "+nombre 
                + " y es de " + saldo);
        return true;
    }
    
    /**
     * Se mueve el jugador a la casilla pasada como parametro 
     * si no esta encarcelado.
     * @param numCasilla
     * @return TRUE si se ha movido, FALSE si no.
     */
    boolean moverACasilla(int numCasilla) {
        if(isEncarcelado()) {
            return false;
        }else{
            numCasillaActual = numCasilla;
            puedeComprar = false;
            Diario.getInstance().ocurreEvento("El jugador " + nombre + 
                    " se mueve a la casilla " + 
                    Float.toString(numCasillaActual));
            return true;
        }
    }
    
    /**
     * El jugador obtiene un salvoconducto.
     * @param sorpresa
     * @return TRUE si obtiene salvoconducto, FALSE si no.
     */
    boolean obtenerSalvoconducto(SorpresaSalirCarcel sorpresa) {
        if(encarcelado) {
            return false;
        }else{
            salvoconducto = sorpresa;
            return true;
        }
    }
    
    /**
     * El jugador paga una cantidad y se le modifica el saldo.
     * @param cantidad
     * @return TRUE si se ha modificado el saldo, FALSE si no.
     */
    boolean paga(float cantidad) {
        return modificarSaldo(cantidad * (-1));
    }
    
    /**
     * El jugador paga el alquiler si no esta encarcelado.
     * @param cantidad
     * @return TRUE si lo paga, FALSE si no.
     */
    boolean pagaAlquiler(float cantidad) {
        if(isEncarcelado()) {
            return false;
        }else{
            return paga(cantidad);
        }
    }

    /**
     * El jugador paga el impuesto si no esta encarcelado.
     * @param cantidad
     * @return TRUE si lo paga, FALSE si no.
     */
    boolean pagaImpuesto(float cantidad) {
        if(isEncarcelado()) {
            return false;
        }else{
            return paga(cantidad);
        }
    }
    
    /**
     * Comprueba si el jugador pasa por la salida y se le cobra.
     * @return siempre TRUE.
     */
    boolean pasaPorSalida() {
        modificarSaldo(PasoPorSalida);
        Diario.getInstance().ocurreEvento("El jugador " + nombre + " recibe " + 
                Float.toString(PasoPorSalida) + " por pasar por la salida");
        return true;
    }
    
    /**
     * El jugador pierde salvoconducto.
     */
    private void perderSalvoconducto() {
        salvoconducto.usada();
        salvoconducto = null;
    }
    
    /**
     * Consulta si el jugador puede comprar una casilla.
     * @return TRUE si puede comprar, FALSE si no.
     */
    boolean puedeComprarCasilla() {
        puedeComprar = !encarcelado;
        return puedeComprar;
    }
    
    /**
     * Comprueba si puede salir de la carcel pagando.
     * @return TRUE si puede salir pagando, FALSE si no.
     */
    private boolean puedoSalirCarcelPagando() {
        return saldo >= PrecioLibertad;
    }
    
    /**
     * Comprueba si se puede edificar una casa.
     * @param propiedad
     * @return TRUE si se puede, FALSE si no.
     */
    private boolean puedoEdificarCasa(TituloPropiedad propiedad) {
        if(encarcelado) {
            return false;
        }
        if(propiedad.getNumCasas() <= CasasMax) {
            return saldo >= propiedad.getPrecioEdificar();
        }else{
            return false;
        }
    }
    
    /**
     * Comprueba si se puede edificar un hotel.
     * @param propiedad
     * @return TRUE si se puede, FALSE si no.
     */
    private boolean puedoEdificarHotel(TituloPropiedad propiedad) {
        if(encarcelado) {
            return false;
        }
        if((propiedad.getNumHoteles() <= HotelesMax) && 
                (propiedad.getNumCasas() == CasasMax)) {
            return saldo >= propiedad.getPrecioEdificar();
        }else{
            return false;
        }
    }
    
    /**
     * Consulta si el jugador puede gastarse ese dinero.
     * @param precio
     * @return TRUE si puede gastarlo, FALSE si no.
     */
    protected boolean puedoGastar(float precio) {
        if(isEncarcelado()) {
            return false;
        }else{
            return saldo >= precio;
        }
    }
    
    /**
     * EL jugador recibe una cantidad de dinero. si no esta encarcelado.
     * @param cantidad
     * @return TRUE si recibe, FALSE si no.
     */
    boolean recibe(float cantidad) {
        if(isEncarcelado()) {
            return false;
        }else{
            return modificarSaldo(cantidad);
        }
    }
    
    /**
     * Comprueba si puede salir de la carcel pagando si esta encarcelado.
     * Informa si el saldo del jugador es mayor al precio de salir de la carcel.
     * @return TRUE si puede salir, FALSE si no.
     */
    boolean salirCarcelPagando() {
        if(encarcelado&& puedoSalirCarcelPagando()) {
            paga(PrecioLibertad);
            encarcelado = false;
            Diario.getInstance().ocurreEvento("El jugador " + nombre + 
                    " ha pagado " + Float.toString(PrecioLibertad) + 
                    " y puede salir de la carcel");
            return true;
        }else{
            return false;
        }
    }
    
    /**
     * Comprueba si se puede salir de la carcel tirando.
     * @return TRUE si puede salir, FALSE si no.
     */
    boolean salirCarcelTirando() {
        if(Dado.getInstance().salgoDeLaCarcel()) {
            encarcelado = false;
            Diario.getInstance().ocurreEvento("El jugador " + nombre + 
                    " sale de la carcel gracias a los dados");
            return true;
        }else{
            return false;
        }
    }
    
    /**
     * Comprueba si tiene que gestionar alguna casa u hotel.
     * @return TRUE si tiene casas u hoteles, FALSE si no tiene.
     */
    boolean tieneAlgoQueGestionar() {
        return propiedades != null;
    }
    
    /**
     * Comprueba si tiene un salvoconducto.
     * @return TRUE si tiene, FALSE si no.
     */
    boolean tieneSalvoconducto() {
        return salvoconducto!=null;
    }
    /**
     * Proporciona una informacion del jugador con una cadena de caracteres.
     * @return salida
     */
    @Override 
    public String toString() {
        String todas = "", aux;
        
        for(int i = 0; i < propiedades.size(); i++) {
            aux = propiedades.get(i).toString();
            todas = todas + " ";
            todas = todas + aux;
        }
        String salida = "Jugador: " + getNombre() +
                "\nCasilla Actual: " + numCasillaActual +
                "\nSaldo: " + saldo + 
                "\nNumero de propiedades: " + propiedades.size() + 
                "\nPropiedades: \n" + todas;
      
        return salida;
    }
    
    /**
     * Se vende la propiedad (ip).
     * @param ip indice de la propiedad.
     * @return TRUE si se vende, FALSE si no.
     */
    boolean vender(int ip) {
        if(isEncarcelado()) {
            return false;
        }else{
            if(existeLaPropiedad(ip)) {
                if(propiedades.get(ip).vender(this)) {
                    Diario.getInstance().ocurreEvento("Se ha vendido la "
                            + "propiedad" + propiedades.get(ip).getNombre() +
                            " del jugador " + this.nombre);
                    propiedades.remove(ip);
                    return true;
                }else{
                    return false;
                }
            }else{
                return false;
            }
        }
    }
}

File: civitasRuby/lib/jugador.rb
# encoding:utf-8
module Civitas
  
  
  class Jugador
    
    include Comparable

    attr_reader :nombre,:num_casilla_actual,:casas_max,:casas_por_hotel,
      :propiedades,:puede_comprar,:saldo,:encarcelado,:salvoconducto,
      :especulador

    def initialize(nombre)
      @@casas_max = 4
      @@casas_por_hotel = 4
      @encarcelado = false
      @@hoteles_max = 4
      @nombre = nombre
      @num_casilla_actual = 0
      @@paso_por_salida = 1000
      @@precio_libertad = 200
      @puede_comprar = true
      @saldo = 7500
      @@saldo_inicial = 7500
      @salvoconducto = nil
      @propiedades = Array.new
      @especulador = false
    end
    
    def jugador_copia(otro)
      @nombre = otro.nombre
      @encarcelado = otro.encarcelado
      @num_casilla_actual = otro.numCasillaActual
      @puede_comprar = otro.puedeComprar
      @saldo = otro.saldo
      @salvoconducto = otro.salvoconducto
      @propiedades = otro.propiedades
      for i in @propiedades
        i.actualiza_propietario_por_conversion(self)  
      end
      @especulador = otro.especulador
    end
    
    def self.jugador(otro)
      @nombre = otro.nombre
      @encarcelado = otro.encarcelado
      @num_casilla_actual = otro.numCasillaActual
      @puede_comprar = otro.puedeComprar
      @saldo = otro.saldo
      @salvoconducto = otro.salvoconducto
      @propiedades = otro.propiedades
    end
    
    def cancelar_hipoteca(ip)
        result = false
      if @encarcelado
        return result
      end
      if existe_la_propiedad(ip)
        propiedad = @propiedades.at(ip)
        cantidad = propiedad.get_importe_cancelar_hipoteca
        puedo_gastar = puedo_gastar(cantidad)
        if puedo_gastar
          result = propiedad.cancelar_hipoteca(self)
          if result
            Diario.instance.ocurre_evento("El jugador #{@nombre}" + 
                "cancela la hipoteca de la propiedad #{ip.to_s}")
          end
        end
      end
      return result
    end
    
    def cantidad_casas_hoteles
      cantidad = 0
      for i in (0..@propiedades.size - 1)
        cantidad += @propiedades.at(i).num_casas
        cantidad += @propiedades.at(i).num_hoteles
      end
      return cantidad
    end
    
    def compare_to(otro)
      return @saldo <=> otro.get_saldo
    end
    
    def comprar(titulo)
      result = false
      if @encarcelado
        return result
      end
      if @puede_comprar
        precio = titulo.precio_compra
        if puedo_gastar(precio)
          result = titulo.comprar(self)
          if result
            @propiedades << titulo
            Diario.instance.ocurre_evento("El jugador #{@nombre}" + 
                "comprala propiedad #{titulo.to_string}")
          end
          @puede_comprar = false
        end
      end
      return result
    end
    
    def construir_casa(ip)
      result = false
      if @encarcelado
        return result
      else
        existe = existe_la_propiedad(ip)
        if existe
          propiedad = @propiedades.at(ip)
          puedo_edificar_casa = puedo_edificar_casa(propiedad)
          if puedo_edificar_casa
            result = propiedad.construir_casa(self)
            if result
              Diario.instance.ocurre_evento("El jugador #{@nombre}" +
                  "construye una casa en la propiedad #{ip.to_s}")
            end
          end 
        end
      end
      return result
    end
    
    def construir_hotel(ip)
      result = false
      if @encarcelado
        return result
      end
      if existe_la_propiedad(ip)
        propiedad = @propiedades.at(ip)
        puedo_edificar_hotel = puedo_edificar_hotel(propiedad)
        precio = propiedad.precio_edificar
        if puedo_gastar(precio) && propiedad.num_hoteles < @@hoteles_max && 
            propiedad.num_casas >= @@casas_por_hotel
          puedo_edificar_hotel = true
        end
        if puedo_edificar_hotel
          result = propiedad.construir_hotel(self)
          propiedad.derruir_casa(@@casas_por_hotel, self)
          Diario.instance.ocurre_evento("El jugador #{@nombre}" + 
              "construye un hotel en la propiedad #{ip.to_s}")
        end
      end
      return result
    end
    
    def debe_ser_encarcelado
      if @encarcelado
        return false
      else
        if !tiene_salvoconducto
          return true
        else
          perder_salvoconducto
          Diario.instance.ocurre_evento("El jugador #{@nombre} se 
libra de la carcel")
          return false
        end
      end
    end
    
    def en_bancarrota
      return @saldo <= 0
    end
    
    def encarcelar(num_casilla_carcel)
      if debe_ser_encarcelado
        mover_a_casilla(num_casilla_carcel)
        @encarcelado = true
        Diario.instance.ocurre_evento("Se ha encarcelado al jugador #{@nombre}")
        return @encarcelado
      end
    end
    
    def get_casas_por_hotel
      return @@casas_por_hotel
    end
    
    def get_especulador
      return @especulador
    end
    
    def get_nombre
      return @nombre
    end
    
    def get_num_casilla_actual
      return @num_casilla_actual
    end
    
    def get_propiedades
      nombres =  Array.new
      for i in 0..@propiedades.size
        nombres << @propiedades.at(i).nombre
      end
      return nombres
    end
    
    def get_puede_comprar
      return @puede_comprar
    end
    
    def get_saldo
      return @saldo
    end
    
    def get_salvoconducto
      return @salvoconducto
    end
    
    def hipotecar(ip)
      result = false
      if @encarcelado
        return result
      end
      if existe_la_propiedad(ip)
        propiedad = @propiedades.at(ip)
        result= propiedad.hipotecar(self)
      end
      if result
        Diario.instance.ocurre_evento("El jugador #{@nombre} hipoteca 
la propiedad #{ip.to_s}")
      end
      return result
    end
    
    def is_encarcelado
      return @encarcelado
    end
    
    def modificar_saldo(cantidad)
      @saldo += cantidad
      Diario.instance.ocurre_evento("Se ha modificado el saldo." +
          "\nNuevo saldo: #{@saldo}")
    end
    
    def mover_a_casilla(num_casilla)
      if @encarcelado
        return false
      else
        @num_casilla_actual = num_casilla
        @puede_comprar = false
        Diario.instance.ocurre_evento("El juagdor #{@nombre} se ha movido a 
la casilla #{@num_casilla_actual.to_s}")
        return true
      end
    end
    
    def obtener_salvoconducto(sorpresa)
      if @encarcelado
        return false
      else
        @salvoconducto = sorpresa
        return true
      end
    end
    
    def paga(cantidad)
      paga = modificar_saldo(cantidad * -1)
      return paga
    end
    
    def paga_alquiler(cantidad)
      if @encarcelado
        return false
      else
        paga = paga(cantidad)
        return paga
      end
      
    end
    
    def paga_impuesto(cantidad)
      if @encarcelado
        return false
      else
        paga = paga(cantidad)
        return paga
      end
    end
    
    def pasa_por_salida
      modificar_saldo(@@paso_por_salida)
      Diario.instance.ocurre_evento("El jugador #{@nombre} ha recibido 
#{@@paso_por_salida.to_s} por pasar po la Salida.")
    end
    
    def puede_comprar_casilla
      @puede_comprar = !encarcelado
      return @puedo_comprar
    end
    
    def recibe(cantidad)
      if @encarcelado
        return false
      else
        recibe = modificar_saldo(cantidad)
        return recibe
      end
    end
    
    def salir_carcel_pagando
      if @encarcelado && puede_salir_carcel_pagando
        paga(@@precio_libertad)
        @encarcelado = false
        Diario.instance.ocurre_evento("El jugador #{@nombre} ha pagado 
#{@@precio_libertad} y sale de la carcel.")
        return true
      else 
        return false
      end
    end
    
    def salir_carcel_tirando
      if Dado.instance.salgo_de_la_carcel
        @encarcelado = false
        Diario.instance.ocurre_evento("El jugador #{@nombre} sale de la carcel 
gracias a los dados.")
        return true
      else 
        return false
      end
    end
    
    def tiene_algo_que_gestionar
      return @propiedades != nil

    end
    
    def tiene_salvoconducto
      return @salvoconducto != nil
    end
    
    def to_string
      for i in 0...@propiedades.size
        todas = todas + @propiedades.at(i).nombre + " "
      end
      salida = "Nombre: " + @nombre + "\nSaldo: " + @saldo.to_s + 
        "\nCasilla actual: " + @num_casilla_actual.to_s + 
        "\n¿Esta encarcelado? " + @encarcelado.to_s +
        "\nNumero de propiedades: " + @propiedades.size.to_s + 
        "\nPropiedades:\n" + todas
      return salida
    end
    
    def vender(ip)
      if @encarcelado
        return false
      else
        if existe_la_propiedad(ip)
          vendida = @propiedades.at(ip).vender(self)
          if vendida
            Diario.instance.ocurre_evento("La propiedad 
#{@propiedades.at(ip).nombre} ha sido vendida por #{@nombre}")
            @propiedades.delete_at(ip)
            return true
          else
            return false
          end
        else 
          return false
        end
      end
    end
    
    def get_nombre_propiedades
      nombres = Array.new
      for i in 0...@propiedades.size
        nombres << @propiedades[i].nombre
      end
      return nombres
    end
    
    
    private
    
    def existe_la_propiedad(ip)
      return ip <= @propiedades.size
    end
    
    def get_casas_max
      return @@casas_max
    end
    
    def get_hoteles_max
      return @@hoteles_max
    end
    
    def get_precio_libertad
      return @@precio_libertad
    end
    
    def get_premio_paso_salida
      return @@paso_por_salida
    end
    
    def perder_salvoconducto
      @salvoconducto.usada
      @salvoconducto = nil
    end
    
    def puede_salir_carcel_pagando
      return @saldo>=@@PrecioLibertad
    end
    
    def puedo_edificar_casa(propiedad)
      if @encarcelado
        return false
      else
        if propiedad.num_casas <= @@casas_max
          return @saldo >= propiedad.precio_edificar
        else
          return false
        end
      end
    end
    
    def puedo_edificar_hotel(propiedad)
      if @encarcelado
        return false
      else
        if (propiedad.num_hoteles <= @@hoteles_max) && (propiedad.num_casas ==4)
          return @saldo >= propiedad.precio_edificar
        else
          return false
        end
      end
    end
    
    def puedo_gastar(precio)
      if @encarcelado
        return false
      else
        return @saldo >= precio
      end
    end
    
    
  end
end